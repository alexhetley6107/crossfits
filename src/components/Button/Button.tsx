import React from 'react';
import { theme } from '@/themes/theme';
import { ButtonVariant, SizeUnion } from '@/types';
import styled from 'styled-components';
import { Loader } from '../Loader';

export type ButtonProps = {
  className?: string;
  text: string;
  disabled?: boolean;
  size?: SizeUnion;
  variant?: ButtonVariant;
  fullWidth?: boolean;
  loading?: boolean;
  onClick?: () => void;
};

const sizeMap: Record<string, Record<SizeUnion, string>> = {
  padding: {
    sm: '5px 12px',
    lg: '10px 24px',
  },
  height: {
    sm: '28px',
    lg: '40px',
  },
  font: {
    sm: '14px',
    lg: '18px',
  },
  loader: {
    sm: '20',
    lg: '24',
  },
};
const variantMap: Record<string, Record<ButtonVariant, string>> = {
  color: {
    filled: theme.colors.white,
    outlined: theme.colors.primary,
  },
  background: {
    filled: theme.colors.primary,
    outlined: theme.colors.white,
  },
  loader: {
    filled: theme.colors.white,
    outlined: theme.colors.primary,
  },
};

const StyledButton = styled.button<{
  $size: SizeUnion;
  $variant: ButtonVariant;
  $fullWidth?: boolean;
  disabled?: boolean;
}>`
  position: relative;
  outline: none;
  border-radius: ${theme.radius.R6};
  cursor: pointer;
  transition: all 0.1s;
  border: 1px solid ${theme.colors.primary};
  width: ${({ $fullWidth }) => ($fullWidth ? '100%' : 'auto')};
  opacity: ${({ disabled }) => (disabled ? 0.4 : 1)};
  padding: ${({ $size }) => sizeMap.padding[$size]};
  height: ${({ $size }) => sizeMap.height[$size]};
  font-size: ${({ $size }) => sizeMap.font[$size]};
  color: ${({ $variant }) => variantMap.color[$variant]};
  background-color: ${({ $variant }) => variantMap.background[$variant]};
  &:focus ~ .focus {
    opacity: 1;
  }
`;

const LoaderWrapper = styled.div`
  position: absolute;
  inset: 1px;
  display: flex;
  justify-content: center;
  align-items: center;
  background-color: inherit;
  border-radius: ${theme.radius.R6};
`;

export const Button: React.FC<ButtonProps> = ({
  text,
  disabled,
  onClick,
  size = 'sm',
  variant = 'filled',
  fullWidth = false,
  className,
  loading,
  ...props
}) => {
  return (
    <div className={className}>
      <StyledButton
        data-testid="button"
        $size={size}
        $variant={variant}
        $fullWidth={fullWidth}
        onClick={onClick}
        disabled={disabled}
        {...props}
      >
        {text}

        {loading && (
          <LoaderWrapper>
            <Loader size={+sizeMap.loader[size]} color={variantMap.loader[variant]} />
          </LoaderWrapper>
        )}
      </StyledButton>
    </div>
  );
};
