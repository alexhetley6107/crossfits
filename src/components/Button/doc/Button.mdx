import { Canvas, Meta } from '@storybook/blocks';
import { Example } from './Example.tsx';
import * as Button from './Button.stories.tsx';

<Meta of={Button} title="Button" />

# Button

<Canvas of={Button.Base} />

## Usage

```ts
import { Button } from 'remindee-ui-kit';

export const Example = () => {
  return <Button text="Button" />;
};
```

### Arguments

- **text** - A string value of the text inside button.
- **disabled** - A boolean indicating whether the button should be disabled or not.
- **fullWidth** - A boolean value indicating whether the button should take all available width.
- **size** - A string value which determine button's size layout. Options: `sm` and `lg`.
- **variant** - A string value which determine button's display layout. Options: `filled` and `outlined`.
