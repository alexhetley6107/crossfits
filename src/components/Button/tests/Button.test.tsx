import { render, screen } from '@testing-library/react';
import { describe, expect, it } from 'vitest';
import { Button } from '../Button';
import '@testing-library/jest-dom';
import 'jest-styled-components';

describe('Button component', () => {
  describe('General', () => {
    it('Render', () => {
      render(<Button text="Button" />);
      const button = screen.getByTestId('button');
      expect(button).toBeInTheDocument();
    });
  });
  describe('Props', () => {
    it('text', () => {
      render(<Button text="Button" />);
      const text = screen.getByText('Button');
      expect(text).toBeInTheDocument();
    });
    describe('Size', () => {
      it('Small', () => {
        render(<Button text="Button" size="sm" />);
        const button = screen.getByTestId('button');
        expect(button).toMatchSnapshot();
      });
      it('Large', () => {
        render(<Button text="Button" size="lg" />);
        const button = screen.getByTestId('button');
        expect(button).toMatchSnapshot();
      });
    });
    describe('Variant', () => {
      it('Filled', () => {
        render(<Button text="Button" variant="filled" />);
        const button = screen.getByTestId('button');
        expect(button).toMatchSnapshot();
      });
      it('Outlined', () => {
        render(<Button text="Button" variant="outlined" />);
        const button = screen.getByTestId('button');
        expect(button).toMatchSnapshot();
      });
    });
    it('Full Width', () => {
      render(<Button text="Button" fullWidth />);
      const button = screen.getByTestId('button');
      expect(button).toMatchSnapshot();
    });
    it('Disabled', () => {
      render(<Button text="Button" disabled />);
      const button = screen.getByTestId('button');
      expect(button).toMatchSnapshot();
    });
    describe('Loading', () => {
      it('Filled', () => {
        render(<Button text="Button" variant="filled" loading />);
        const button = screen.getByTestId('button');
        expect(button).toMatchSnapshot();
      });
      it('Outlined', () => {
        render(<Button text="Button" variant="outlined" loading />);
        const button = screen.getByTestId('button');
        expect(button).toMatchSnapshot();
      });
    });
  });
});
