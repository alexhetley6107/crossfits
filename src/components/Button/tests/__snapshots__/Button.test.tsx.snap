// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button component > Props > Disabled 1`] = `
.c0 {
  border-radius: 5px;
  cursor: pointer;
  transition: all 0.1s;
  border: 1px solid #346beb;
  width: auto;
  opacity: 0.4;
  padding: 5px 12px;
  font-size: 14px;
  color: #fff;
  background-color: #346beb;
}

.c0:hover:not([disabled]) {
  transform: translateY(-1px);
}

.c0:active:not([disabled]) {
  transform: translateY(1px);
}

<button
  class="c0"
  data-testid="button"
  disabled=""
>
  Button
</button>
`;

exports[`Button component > Props > Full Width 1`] = `
.c0 {
  border-radius: 5px;
  cursor: pointer;
  transition: all 0.1s;
  border: 1px solid #346beb;
  width: 100%;
  opacity: 1;
  padding: 5px 12px;
  font-size: 14px;
  color: #fff;
  background-color: #346beb;
}

.c0:hover:not([disabled]) {
  transform: translateY(-1px);
}

.c0:active:not([disabled]) {
  transform: translateY(1px);
}

<button
  class="c0"
  data-testid="button"
>
  Button
</button>
`;

exports[`Button component > Props > Size > Large 1`] = `
.c0 {
  border-radius: 5px;
  cursor: pointer;
  transition: all 0.1s;
  border: 1px solid #346beb;
  width: auto;
  opacity: 1;
  padding: 10px 24px;
  font-size: 18px;
  color: #fff;
  background-color: #346beb;
}

.c0:hover:not([disabled]) {
  transform: translateY(-1px);
}

.c0:active:not([disabled]) {
  transform: translateY(1px);
}

<button
  class="c0"
  data-testid="button"
>
  Button
</button>
`;

exports[`Button component > Props > Size > Small 1`] = `
.c0 {
  border-radius: 5px;
  cursor: pointer;
  transition: all 0.1s;
  border: 1px solid #346beb;
  width: auto;
  opacity: 1;
  padding: 5px 12px;
  font-size: 14px;
  color: #fff;
  background-color: #346beb;
}

.c0:hover:not([disabled]) {
  transform: translateY(-1px);
}

.c0:active:not([disabled]) {
  transform: translateY(1px);
}

<button
  class="c0"
  data-testid="button"
>
  Button
</button>
`;

exports[`Button component > Props > variant > Filled 1`] = `
.c0 {
  border-radius: 5px;
  cursor: pointer;
  transition: all 0.1s;
  border: 1px solid #346beb;
  width: auto;
  opacity: 1;
  padding: 5px 12px;
  font-size: 14px;
  color: #fff;
  background-color: #346beb;
}

.c0:hover:not([disabled]) {
  transform: translateY(-1px);
}

.c0:active:not([disabled]) {
  transform: translateY(1px);
}

<button
  class="c0"
  data-testid="button"
>
  Button
</button>
`;

exports[`Button component > Props > variant > Outlined 1`] = `
.c0 {
  border-radius: 5px;
  cursor: pointer;
  transition: all 0.1s;
  border: 1px solid #346beb;
  width: auto;
  opacity: 1;
  padding: 5px 12px;
  font-size: 14px;
  color: #346beb;
  background-color: #fff;
}

.c0:hover:not([disabled]) {
  transform: translateY(-1px);
}

.c0:active:not([disabled]) {
  transform: translateY(1px);
}

<button
  class="c0"
  data-testid="button"
>
  Button
</button>
`;
